static/src/javascripts/projects/common/modules/spacefinder.ts

```diff
@@ -0,0 +1,448 @@
+	);
+	exclusions.absoluteMinAbove = result.exclusions;
+	candidates = result.filtered;
+
+	// enforce minAbove and minBelow rules
+	result = partitionCandidates(candidates, (candidate) => {
+		const farEnoughFromTopOfBody = candidate.top >= rules.minAbove;
+		const farEnoughFromBottomOfBody =
+			candidate.top + rules.minBelow <= measurements.bodyHeight;
+		return farEnoughFromTopOfBody && farEnoughFromBottomOfBody;
+	});
+	exclusions.aboveAndBelow = result.exclusions;
+	candidates = result.filtered;
+
+	// enforce content meta rule
+	if (rules.clearContentMeta !== undefined) {
+		result = partitionCandidates(
+			candidates,
+			(candidate) =>
+				!!measurements.contentMeta &&
+				candidate.top >
+					measurements.contentMeta.bottom +
+						(rules.clearContentMeta ?? 0), // why do we need to do this despite the type guard?
+		);
+		exclusions.contentMeta = result.exclusions;
+		candidates = result.filtered;
+	}
+
+	// enforce selector rules
+	if (rules.selectors) {
+		const selectorExclusions: SpacefinderItem[] = [];
+		for (const [selector, rule] of Object.entries(rules.selectors)) {
+			result = partitionCandidates(candidates, (candidate) =>
+				testCandidates(
+					rule,
+					candidate,
+					measurements.opponents
+						? measurements.opponents[selector]
+						: [],
+				),
+			);
+			exclusions[selector] = result.exclusions;
+			selectorExclusions.push(...result.exclusions);
+		}
+
+		candidates = candidates.filter(
+			(candidate) => !selectorExclusions.includes(candidate),
+		);
+	}
+
+	if (rules.filter) {
+		result = partitionCandidates(candidates, rules.filter);
+		exclusions.custom = result.exclusions;
+		candidates = result.filtered;
+	}
+
+	return candidates;
+};
+
+class SpaceError extends Error {
+	constructor(rules: SpacefinderRules) {
+		super();
+		this.name = 'SpaceError';
+		this.message = `There is no space left matching rules from ${rules.bodySelector}`;
+	}
+}
+/**
+ * Wait for the page to be ready (images loaded, rich links upgraded, interactives loaded)
+ * or for LOADING_TIMEOUT to elapse, whichever comes first.
+ * @param  {SpacefinderRules} rules
+ * @param  {SpacefinderOptions} options
+ */
+const getReady = (rules: SpacefinderRules, options: SpacefinderOptions) =>
+	Promise.race([
+		new Promise(() => {
+			window.setTimeout(noop, LOADING_TIMEOUT);
+		}),
+		Promise.all([
+			options.waitForImages ? onImagesLoaded(rules) : true,
+			options.waitForLinks ? onRichLinksUpgraded(rules) : true,
+			options.waitForInteractives ? onInteractivesLoaded(rules) : true,
+		]),
+	]);
+
+const getCandidates = (
+	rules: SpacefinderRules,
+	exclusions: SpacefinderExclusions,
+) => {
+	let candidates = query(rules.bodySelector + rules.slotSelector);
+	let result:
+		| {
+				filtered: HTMLElement[];
+				exclusions: HTMLElement[];
+		  }
+		| undefined;
+	if (rules.fromBottom) {
+		candidates.reverse();
+	}
+	if (rules.startAt) {
+		let drop = true;
+		result = partitionCandidates(candidates, (candidate) => {
```

Comment:
Would also work [with the partition method from libs](https://github.com/guardian/frontend/blob/93f6b71efba697735006e32c70e1b9ee5874cb6e/static/src/javascripts/lib/partition.ts).

Reply:
I've combined both of these suggestions into:
```
let [filtered, exclusions] = partition()
```
(I think it needs to be a `let` since the variables are getting reassigned further down)